timeout_handler.go

```diff
@@ -0,0 +1,231 @@
+		log:   log,
+	}
+
+	go t.run()
+
+	return t
+}
+
+func (t *TimeoutHandler) GetTime() time.Time {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	return t.now
+}
+
+func (t *TimeoutHandler) run() {
+	for t.shouldRun() {
+		select {
+		case now := <-t.ticks:
+			t.lock.Lock()
+			t.now = now
+			t.lock.Unlock()
+
+			t.maybeRunTasks()
+		case <-t.close:
+			return
+		}
+	}
+}
+
+func (t *TimeoutHandler) maybeRunTasks() {
+	// go through the heap executing relevant tasks
+	for {
+		t.lock.Lock()
+		if t.heap.Len() == 0 {
+			t.lock.Unlock()
+			break
+		}
+
+		next := t.heap[0]
+		if next.Deadline.After(t.now) {
+			t.lock.Unlock()
+			break
+		}
+
+		heap.Pop(&t.heap)
+		delete(t.tasks[string(next.NodeID)], next.TaskID)
+		t.lock.Unlock()
+		t.log.Debug("Executing timeout task", zap.String("taskid", next.TaskID))
+		next.Task()
+	}
+}
+
+func (t *TimeoutHandler) shouldRun() bool {
+	select {
+	case <-t.close:
+		return false
+	default:
+		return true
+	}
+}
+
+func (t *TimeoutHandler) Tick(now time.Time) {
+	select {
+	case t.ticks <- now:
+	default:
+		t.log.Debug("Dropping tick in timeouthandler")
+	}
+}
+
+func (t *TimeoutHandler) AddTask(task *TimeoutTask) {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	if _, ok := t.tasks[string(task.NodeID)]; !ok {
+		t.log.Info("Attempting to add a task for an unknown node", zap.Stringer("node", task.NodeID))
+	}
+
+	// adds a task to the heap and the tasks map
+	if _, ok := t.tasks[string(task.NodeID)][task.TaskID]; ok {
+		t.log.Debug("Trying to add an already included task", zap.Stringer("node", task.NodeID), zap.String("Task ID", task.TaskID))
+		return
+	}
+
+	t.tasks[string(task.NodeID)][task.TaskID] = task
+	t.log.Debug("Adding timeout task", zap.Stringer("node", task.NodeID), zap.String("taskid", task.TaskID))
+	heap.Push(&t.heap, task)
+}
+
+func (t *TimeoutHandler) RemoveTask(nodeID NodeID, ID string) {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	if _, ok := t.tasks[string(nodeID)]; !ok {
+		t.log.Info("Attempting to add a remove a task for an unknown node", zap.Stringer("node", nodeID))
+		return
+	}
+
+	if _, ok := t.tasks[string(nodeID)][ID]; !ok {
+		return
+	}
+
+	// find the task using the task map
+	// remove it from the heap using the index
+	t.log.Debug("Removing timeout task", zap.String("taskid", ID))
+	heap.Remove(&t.heap, t.tasks[string(nodeID)][ID].index)
+	delete(t.tasks[string(nodeID)], ID)
+}
+
+func (t *TimeoutHandler) Close() {
+	select {
+	case <-t.close:
+		return
+	default:
+		close(t.close)
+	}
+}
+
+// FindTask returns the first TimeoutTask assigned to [node] that contains any sequence in [seqs].
+// A sequence is considered "contained" if it falls between a task's Start (inclusive) and End (inclusive).
+func (t *TimeoutHandler) FindTask(node NodeID, seqs []uint64) *TimeoutTask {
+	var task *TimeoutTask
+seq:
+	for _, seq := range seqs {
+		for _, t := range t.tasks[string(node)] {
+			if seq >= t.Start && seq <= t.End {
+				task = t
+				break seq
```

Comment:
can we just return `t` and return `nil` in line 185 and remove the `seq:` label? Makes things more explicit, and I personally just hate goto labels, maybe that's just me...